### ISSUE TITLE
Support including favorite/bookmark value in the /posts api

### ISSUE THREAD

**Issue by justinctlam** [SoftwareEngineeringDaily/software-engineering-daily-api#83]
Is it possible to get back the favorite/bookmark status of the podcast/post when calling the /posts api? Seems a bit inefficient to have to make another /favorites call after getting the /posts items. Thoughts?

**Comment by jasonify** [comment]
👍  Yeah, and it should be pretty similar to how we do it for votes.

**Comment by joshmorel** [comment]
I'll give this a stab. @jasonify  feel free to assign me.

**Comment by joshmorel** [comment]
I'm working on implementing this and trying to find the best way to integrate this into the post model list method, which already has a lot going on. 

I'm wondering if instead of having the favorite.model we use the [approach recommended here](http://seanhess.github.io/2012/02/01/mongodb_relational.html) with favorite relationship in both the post & user document. I come from an RDBMS background where a junction table is the best (and only) option, but with a document DB I doesn't seem like it is. 

This way for join queries, instead of a lot of code, we could use built-in moongoose methods and push a single query to the DB instead of re-querying it multiple times. 

Let me know what you think. We can always try first with favorites which isn't used by any clients then see how it could be apply to votes.

Also, @TheHollidayInn , I'd love your input here!

**Comment by joshmorel** [comment]
I have a WIP branch to show how this would work. So far I've updated the user & posts model to have favoritedPosts & favoritedBy reference arrays respectively. I updated posts/{id}/favorite & posts/{id}/unfavorite paths to show how that part would work.

It's a bit simpler IMHO but that won't be the major benefit. That will come with querying by user or post - it should be a lot simpler than the separate favorite collection. But I'll add that if there's agreement this is the right path. 

Only manual tests so far, so there's some more rigour to be applied as well as these additional items to work on/consider:

* [ ] Add favorite status with each post returned for current user (original request)
* [ ] GET favorites/ - should this return posts favorited (see issue #95 )
* [ ] Should we consider using bookmark instead of favorite now if that is the go forward standard across the apps?

**Comment by joshmorel** [comment]
This seems like the best I've found on the two options for storing a [many-to-many association](https://github.com/JKHeadley/rest-hapi#many_many-data-storage) with MongoDB/Mongoose.

They favour using separate collection as the default for technical reasons at the expense of human readability... so I'll do more work tomorrow to figure out what is the best for our use case. Sticking with separate collection may be better but there's gotta be a more friendly way to emulate joins.

**Comment by joshmorel** [comment]
Turns out virtuals provides the best of both worlds - it supports MongoDB storage best practices with associative collections while enabling the ease of querying of associative embedded references as can be see in PR.  

We  should consider this for future work and potentially for existing code.

**Comment by crablar** [comment]
@jasonify Justin is blocked on this--could you help him out this coming week?
